File: Wells/Source/VievModels/EditWellViewModel.vb
﻿Imports Wells.YPFModel
Imports Wells.ViewBase
Imports Wells.CorePersistence.Repositories

Public Class EditWellViewModel
    Inherits BaseViewModel

    Property WellName As String
        Get
            Return _WellName
        End Get
        Set
            _WellName = Value
            CType(AcceptCommand, RelayCommand).RaiseCanExecuteChanged()
            CType(EditMeasurementCommand, RelayCommand).RaiseCanExecuteChanged()
            CType(NewMeasurementCommand, RelayCommand).RaiseCanExecuteChanged()
            CType(NewExternalLinkCommand, RelayCommand).RaiseCanExecuteChanged()
        End Set
    End Property

    Property X As Double
    Property Y As Double
    Property Z As Double
    Property Latitude As Double
    Property Longitude As Double
    Property Type As Integer
    Property Height As Double
    Property Exists As Boolean
    Property Bottom As Double
    Property SoilAnalyses As New List(Of SoilAnalysis)
    Property WaterAnalyses As New List(Of WaterAnalysis)
    Property FLNAAnalyses As New List(Of FLNAAnalysis)
    Property Measurements As New List(Of Measurement)
    Property Files As New List(Of ExternalFile)

    ReadOnly Property HasWell As Boolean
    ReadOnly Property NameEditable As Boolean
    ReadOnly Property WellTypes As New List(Of String) From {"Pozo", "Sondeo"}

    Private _well As Well
    Private _repo As WellsRepository
    Private _measurementEditViewModel As EditMeasurementViewModel
    Private _WellName As String
    ' Private _View As IView

    Event CloseDialog(result As Boolean)
    Event MustRebind(target As String)

    'Property View As IView
    '    Get
    '        Return _View
    '    End Get
    '    Set
    '        _View = Value
    '        _measurementEditViewModel.View = _View
    '    End Set
    'End Property

    Sub New()
        MyBase.New(Nothing)
        HasWell = False
        NameEditable = True
        _well = New Well()
        _measurementEditViewModel = New EditMeasurementViewModel(_well)
        _repo = RepositoryWrapper.Instance.Wells
    End Sub

    Sub New(w As Well)
        MyBase.New(Nothing)
        _well = w
        _measurementEditViewModel = New EditMeasurementViewModel(_well)
        HasWell = True
        NameEditable = False
        InitializeWell()
        _repo = RepositoryWrapper.Instance.Wells
    End Sub

    Private Sub InitializeWell()
        WellName = _well.Name
        X = _well.X
        Y = _well.Y
        Z = _well.Z
        Latitude = _well.Latitude
        Longitude = _well.Longitude
        Type = _well.WellType
        Height = _well.Height
        Exists = _well.Exists
        Bottom = _well.Bottom
        SoilAnalyses = _well.SoilAnalyses
        WaterAnalyses = _well.WaterAnalyses
        FLNAAnalyses = _well.FLNAAnalyses
        Measurements = _well.Measurements
        Files = _well.Files

        Measurements.Sort()
        SoilAnalyses.Sort()
        WaterAnalyses.Sort()
        FLNAAnalyses.Sort()
        Files.Sort()
    End Sub

    Private Sub CreateOrEditWell()
        If Not HasWell Then
            _well.Name = WellName
        End If

        _well.X = X
        _well.Y = Y
        _well.Z = Z
        _well.Latitude = Latitude
        _well.Longitude = Longitude
        _well.WellType = Type
        _well.Bottom = Bottom
        _well.Exists = Exists
        _well.Height = Height
        _well.SoilAnalyses = SoilAnalyses
        _well.WaterAnalyses = WaterAnalyses
        _well.FLNAAnalyses = FLNAAnalyses
        _well.Measurements = Measurements
        _well.Files = Files
    End Sub

    Private Function IsValidWell()
        If HasWell Then
            Return True
        Else
            If Not String.IsNullOrEmpty(WellName) Then
                Return Not _repo.ContainsName(WellName)
            End If
        End If
        Return False
    End Function

    Private Sub RemoveAll()

        RepositoryWrapper.Instance.Measurements.RemoveRange(_well.Measurements)
        RepositoryWrapper.Instance.SoilAnalyses.RemoveRange(_well.SoilAnalyses)
        RepositoryWrapper.Instance.WaterAnalyses.RemoveRange(_well.WaterAnalyses)
        RepositoryWrapper.Instance.FLNAAnalyses.RemoveRange(_well.FLNAAnalyses)
        RepositoryWrapper.Instance.ExternalFiles.RemoveRange(_well.Files)
        _repo.Remove(_well)

    End Sub

    Protected Overrides Sub SetValidators()
        Throw New NotImplementedException()
    End Sub

    Protected Overrides Sub SetCommandUpdates()
        Throw New NotImplementedException()
    End Sub

    ReadOnly Property DeleteWellCommand As ICommand = New RelayCommand(Sub()
                                                                           If View.ShowMessageBox("¿Desea eliminar este pozo? Esto borrará toda la información asociada al mismo.", "Borrar pozo") Then
                                                                               RemoveAll()
                                                                               RepositoryWrapper.Instance.SaveChanges()
                                                                               RaiseEvent CloseDialog(True)
                                                                           End If
                                                                       End Sub,
                                                                   Function()
                                                                       Return HasWell
                                                                   End Function,
                                                                   AddressOf OnError)

    ReadOnly Property AcceptCommand As ICommand = New RelayCommand(Sub()
                                                                       CreateOrEditWell()
                                                                       If HasWell Then
                                                                           _repo.Update(_well)
                                                                       Else
                                                                           _repo.Add(_well)
                                                                       End If
                                                                       RepositoryWrapper.Instance.SaveChanges()
                                                                       RaiseEvent CloseDialog(True)
                                                                   End Sub,
                                                                   Function() IsValidWell(),
                                                                   AddressOf OnError)


    Property NewMeasurementCommand As ICommand = New RelayCommand(Sub()
                                                                      Dim result = CType(View, WellEditingDialog).ShowEditMeasurementDialog(_measurementEditViewModel)
                                                                      If result Then
                                                                          Measurements.Sort()
                                                                          RaiseEvent MustRebind(NameOf(Measurements))
                                                                      End If
                                                                  End Sub,
                                                      Function() HasWell, AddressOf OnError)

    Property EditMeasurementCommand As ICommand = New RelayCommand(Sub(param)
                                                                       If param IsNot Nothing AndAlso TypeOf param Is Measurement Then
                                                                           Dim vm As New EditMeasurementViewModel(CType(param, Measurement))
                                                                           Dim result = CType(View, WellEditingDialog).ShowEditMeasurementDialog(vm)
                                                                           If result Then
                                                                               Measurements.Sort()
                                                                               RaiseEvent MustRebind(NameOf(Measurements))
                                                                           End If
                                                                       End If
                                                                   End Sub,
                                                      Function() HasWell, AddressOf OnError)

    Property DeleteMeasurementCommand As ICommand = New RelayCommand(Sub(param)
                                                                         If param IsNot Nothing AndAlso TypeOf param Is Measurement Then
                                                                             Dim measurement = CType(param, Measurement)
                                                                             _well.Measurements.Remove(measurement)
                                                                             RepositoryWrapper.Instance.Measurements.Remove(measurement)
                                                                             Measurements.Remove(measurement)
                                                                             Measurements.Sort()
                                                                             RaiseEvent MustRebind(NameOf(Measurements))
                                                                         End If
                                                                     End Sub,
                                                      Function() HasWell, AddressOf OnError)

    Property NewExternalLinkCommand As ICommand = New RelayCommand(Sub()
                                                                       Dim filename = View.OpenFileDialog("Archivos|*.*", "Elija un archivo")
                                                                       If Not String.IsNullOrEmpty(filename) Then
                                                                           Dim ExternalLink = New ExternalFile(filename) With {.Well = _well}
                                                                           RepositoryWrapper.Instance.ExternalFiles.Add(ExternalLink)
                                                                           Files.Sort()
                                                                           RaiseEvent MustRebind(NameOf(Files))
                                                                       End If
                                                                   End Sub,
                                                      Function() HasWell, AddressOf OnError)

    Property OpenExternalLinkCommand As ICommand = New RelayCommand(Sub(param)
                                                                        If param IsNot Nothing AndAlso TypeOf param Is ExternalFile Then
                                                                            Dim link = CType(param, ExternalFile)
                                                                            link.Open()
                                                                        End If
                                                                    End Sub,
                                                      Function() HasWell, AddressOf OnError)

    Property DeleteExternalLinkCommand As ICommand = New RelayCommand(Sub(param)
                                                                          If param IsNot Nothing AndAlso TypeOf param Is ExternalFile Then
                                                                              If View.ShowMessageBox("Esto eliminará el archivo vinculado a la base de datos. ¿Desea continuar?", "Eliminar archivo") Then
                                                                                  Dim link = CType(param, ExternalFile)
                                                                                  _well.Files.Remove(link)
                                                                                  RepositoryWrapper.Instance.ExternalFiles.Remove(link)
                                                                                  Files.Remove(link)
                                                                                  Files.Sort()
                                                                                  RaiseEvent MustRebind(NameOf(Files))
                                                                              End If
                                                                          End If
                                                                      End Sub,
                                                      Function() HasWell, AddressOf OnError)

End Class

File: DesktopApp/src/ViewModels/EditWellViewModel.cs
using FluentValidation;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Windows.Input;
using Wells.Base;
using Wells.CorePersistence.Repositories;
using Wells.CoreView;
using Wells.CoreView.Validators;
using Wells.CoreView.ViewInterfaces;
using Wells.CoreView.ViewModel;
using Wells.StandardModel.Models;
using Wells.YPFModel;

namespace Wells.View.ViewModels
{
    public class EditWellViewModel : BaseViewModel
    {
        private Well well;
        private string wellName;
        private double x;
        private double y;
        private double z;
        private double latitude;
        private double longitude;
        private int type;
        private double height;
        private bool exists;
        private double bottom;
        private IEditWellView dialog;
        public string WellName { get => wellName; set { SetValue(ref wellName, value); } }


        public double X { get => x; set { SetValue(ref x, value); } }
        public double Y { get => y; set { SetValue(ref y, value); } }
        public double Z { get => z; set { SetValue(ref z, value); } }
        public double Latitude { get => latitude; set { SetValue(ref latitude, value); } }
        public double Longitude { get => longitude; set { SetValue(ref longitude, value); } }
        public int Type { get => type; set { SetValue(ref type, value); } }
        public double Height { get => height; set { SetValue(ref height, value); } }
        public bool Exists { get => exists; set { SetValue(ref exists, value); } }
        public double Bottom { get => bottom; set { SetValue(ref bottom, value); } }
        public ObservableCollection<SoilAnalysis> SoilAnalyses { get; set; }
        public ObservableCollection<WaterAnalysis> WaterAnalyses { get; set; }
        public ObservableCollection<FLNAAnalysis> FLNAAnalyses { get; set; }
        public ObservableCollection<Measurement> Measurements { get; set; }
        public ObservableCollection<ExternalFile> Files { get; set; }

        public bool IsEditing { get; }

        public List<string> WellTypes => Common.EnumDescriptionsToList(typeof(WellType));


        public Well Well => well;


        public EditWellViewModel() : base(null)
        {
            IsEditing = false;
            well = new Well();
            InitializeWell();
        }

        public EditWellViewModel(Well well) : base(null)
        {
            IsEditing = true;
            this.well = well;
            InitializeWell();
    }

        protected override void OnSetView(IView view)
        {
            base.OnSetView(view);
            dialog = (IEditWellView)view;
        }

        void InitializeWell()
        {
            WellName = well.Name;
            X = well.X;
            Y = well.Y;
            Z = well.Z;
            Latitude = well.Latitude;
            Longitude = well.Longitude;
            Type = (int)well.WellType;
            Height = well.Height;
            Exists = well.Exists;
            Bottom = well.Bottom;
            SoilAnalyses = new ObservableCollection<SoilAnalysis>(well.SoilAnalyses.OrderBy(i => i.Date));
            WaterAnalyses = new ObservableCollection<WaterAnalysis>(well.WaterAnalyses.OrderBy(i => i.Date));
            FLNAAnalyses = new ObservableCollection<FLNAAnalysis>(well.FLNAAnalyses.OrderBy(i => i.Date));
            Measurements = new ObservableCollection<Measurement>(well.Measurements.OrderBy(i => i.Date));
            Files = new ObservableCollection<ExternalFile>(well.Files.OrderBy(i => i.CompleteFilename));
        }


        void SaveWell()
        {
            well.Name = WellName;
            well.X = X;
            well.Y = Y;
            well.Z = Z;
            well.Latitude = Latitude;
            well.Longitude = Longitude;
            well.WellType = (WellType)Type;
            well.Bottom = Bottom;
            well.Exists = Exists;
            well.Height = Height;
            well.SoilAnalyses = SoilAnalyses.ToList();
            well.WaterAnalyses = WaterAnalyses.ToList();
            well.FLNAAnalyses = FLNAAnalyses.ToList();
            well.Measurements = Measurements.ToList();
            well.Files = Files.ToList();
        }

        protected override void SetCommandUpdates()
        {
            throw new NotImplementedException();
        }

        protected override void SetValidators()
        {
            Add(nameof(WellName), new List<IValidator>() { new EmptyStringValidator("Nombre"), new WellNameValidator("Nombre") }) ;
        }

        void RemoveAll()
        {
            RepositoryWrapper.Instance.Measurements.RemoveRange(well.Measurements);
            RepositoryWrapper.Instance.SoilAnalyses.RemoveRange(well.SoilAnalyses);
            RepositoryWrapper.Instance.WaterAnalyses.RemoveRange(well.WaterAnalyses);
            RepositoryWrapper.Instance.FLNAAnalyses.RemoveRange(well.FLNAAnalyses);
            RepositoryWrapper.Instance.ExternalFiles.RemoveRange(well.Files);
            RepositoryWrapper.Instance.Wells.Remove(well);
        }


        public ICommand DeleteWellCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (View.ShowYesNoMessageBox("¿Está seguro de eliminar este pozo?", "Eliminar"))
                    {
                        RemoveAll();
                        RepositoryWrapper.Instance.SaveChanges();
                        CloseModalViewCommand.Execute(true);
                    }
                }, (obj) => !IsEditing, OnError);
            }
        }

        public ICommand SaveWellCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    SaveWell();
                    if (IsEditing)
                        RepositoryWrapper.Instance.Wells.Update(well);
                    else
                        RepositoryWrapper.Instance.Wells.Add(well);
                    RepositoryWrapper.Instance.SaveChanges();
                    CloseModalViewCommand.Execute(true);
                }, (obj) => !HasFailures, OnError);
            }
        }

        public ICommand NewMeasurementCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var vm = new EditMeasurementViewModel(well);
                    if (dialog.ShowEditMeasurementDialog(vm))
                    {
                        Measurements.Add(vm.Measurement);
                    }
                }, (obj) => true, OnError);
            }
        }

        public ICommand EditMeasurementCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var vm = new EditMeasurementViewModel(param as Measurement);
                    if (dialog.ShowEditMeasurementDialog(vm))
                    {
                        Measurements.Add(vm.Measurement);
                    }
                }, (obj) => true, OnError);
            }
        }

        public ICommand DeleteMeasurementCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var measurement = (param as Measurement);
                    if (dialog.ShowYesNoMessageBox("¿Desea borrar esta medición?", "Borrar medición"))
                    {
                        Measurements.Remove(measurement);
                    }
                }, (obj) => true, OnError);
            }
        }



        public ICommand NewExternalLinkCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    var filename = View.OpenFileDialog("Archivos|*.*", "Elija un archivo");

                    if (!string.IsNullOrEmpty(filename))
                    {
                        var file = new ExternalFile(filename) { Well = well };
                        Files.Add(file);
                    }
                }, (obj) => true, OnError);
            }
        }

        public ICommand OpenExternalLinkCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if(param != null)
                    {
                        var file = (param as ExternalFile);
                        file.Open();
                    }                   
                }, (obj) => true, OnError);
            }
        }

        public ICommand DeleteExternalLinkCommand
        {
            get
            {
                return new RelayCommand((param) =>
                {
                    if (param != null)
                    {
                        var file = (param as ExternalFile);
                        if (dialog.ShowYesNoMessageBox("¿Desea borrar este archivo?", "Borrar archivo"))
                        {
                            Files.Remove(file);
                        }
                    }
                }, (obj) => true, OnError);
            }
        }
    }
}